File: python/mcp_pager/paginate.py
from __future__ import annotations
import functools
import inspect
import json
from typing import Any, Callable

from mcp.server.fastmcp import FastMCP
from mcp.types import TextContent

from .chunk_store import ChunkStore
from .tokenize import default_token_counter
from .types import (
    ChunkedEvent,
    CursorExpiredEvent,
    PageFetchedEvent,
    PaginateEvent,
    StoreBackend,
)


def paginate(
    mcp: FastMCP,
    *,
    max_tokens: int = 4000,
    ttl_ms: int = 10 * 60 * 1000,
    token_counter: Callable[[str], int] | None = None,
    page_tool_name: str = "get_next_page",
    store: StoreBackend | None = None,
    signing_secret: str | None = None,
    on_paginate: Callable[[PaginateEvent], None] | None = None,
) -> FastMCP:
    """
    Wrap a FastMCP server with token-aware response paging.

    Call once before registering tools. Every tool response that exceeds
    max_tokens is automatically chunked and delivered page by page with
    agent-readable metadata.

    Example:
        mcp = FastMCP("my-server")
        paginate(mcp, max_tokens=4000)

        @mcp.tool()
        async def list_records(limit: int = 500) -> str:
            ...  # large response — automatically paged
    """
    counter = token_counter or default_token_counter
    chunk_store = ChunkStore(ttl_ms, store, signing_secret)

    # Register get_next_page BEFORE patching add_tool so it is never
    # re-paginated by the wrapper.
    @mcp.tool(name=page_tool_name, description="Retrieve the next page of a paginated tool response.")
    async def _get_next_page(cursor: str) -> list[TextContent]:
        result = await chunk_store.get(cursor)
        if result is None:
            _emit(on_paginate, CursorExpiredEvent())
            return [TextContent(
                type="text",
                text="Cursor not found or expired. Please re-invoke the original tool.",
            )]
        _emit(on_paginate, PageFetchedEvent(
            page_index=result["page_index"],
            total_pages=result["total_pages"],
            has_more=result["next_cursor"] is not None,
        ))
        return _build_page_response(
            result["chunk"], result["next_cursor"],
            page_tool_name, result["page_index"], result["total_pages"],
        )

    # Clear outputSchema on get_next_page itself (it returns list[TextContent]
    # which FastMCP still wraps in structured output in 1.27+).
    _clear_output_schema(mcp, page_tool_name)

    # Patch add_tool so every SUBSEQUENT registration gets:
    #   1. A paginating wrapper around the tool function
    #   2. outputSchema cleared (avoids FastMCP structured output validation)
    # add_tool signature: (fn, name=None, title=None, description=None, ...)
    original_add_tool = mcp._tool_manager.add_tool

    def _patched_add_tool(
        fn: Callable, name: str | None = None, **kwargs: Any
    ) -> Any:
        tool_name = name or fn.__name__
        if tool_name == page_tool_name:
            return original_add_tool(fn, name=name, **kwargs)

        is_async = inspect.iscoroutinefunction(fn)

        @functools.wraps(fn)
        async def _wrapped(**call_kwargs: Any) -> list[TextContent]:
            result = (await fn(**call_kwargs)) if is_async else fn(**call_kwargs)
            return await _maybe_paginate(
                result, chunk_store, max_tokens, counter,
                page_tool_name, tool_name, on_paginate,
            )

        registered = original_add_tool(_wrapped, name=name, **kwargs)

        # Clear structured outputSchema AFTER registration so the MCP client
        # won't validate our list[TextContent] responses against it.
        _clear_output_schema(mcp, tool_name)

        return registered

    mcp._tool_manager.add_tool = _patched_add_tool  # type: ignore[method-assign]

    return mcp


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _clear_output_schema(mcp: FastMCP, tool_name: str) -> None:
    """Suppress FastMCP's structured outputSchema for a tool."""
    tool = mcp._tool_manager._tools.get(tool_name)
    if tool and hasattr(tool, "fn_metadata") and tool.fn_metadata is not None:
        tool.fn_metadata.output_schema = None
        tool.fn_metadata.wrap_output = False


async def _maybe_paginate(
    result: Any,
    store: ChunkStore,
    max_tokens: int,
    counter: Callable[[str], int],
    page_tool_name: str,
    tool_name: str,
    on_paginate: Callable[[PaginateEvent], None] | None,
) -> list[TextContent]:
    full_text = _content_to_text(result)
    total_tokens = counter(full_text)

    if total_tokens <= max_tokens:
        # Return as list[TextContent] — consistent with paginated responses.
        return _to_content_list(result)

    chunks = _split_into_chunks(full_text, max_tokens, counter)
    id = await store.save(chunks)

    _emit(on_paginate, ChunkedEvent(
        tool_name=tool_name,
        total_tokens=total_tokens,
        total_chunks=len(chunks),
    ))

    total_pages = len(chunks)
    next_cursor = store.create_cursor(id, 1) if total_pages > 1 else None
    return _build_page_response(chunks[0], next_cursor, page_tool_name, 0, total_pages)


def _build_page_response(
    chunk: str,
    next_cursor: str | None,
    page_tool_name: str,
    page_index: int,
    total_pages: int,
) -> list[TextContent]:
    meta: dict = {
        "hasMore": next_cursor is not None,
        "pageIndex": page_index,
        "totalPages": total_pages,
        "remainingPages": total_pages - page_index - 1,
    }
    if next_cursor:
        meta["nextCursor"] = next_cursor
        meta["instruction"] = (
            f"Call `{page_tool_name}` with nextCursor to get the next page. "
            "Repeat until hasMore is false."
        )
    else:
        meta["instruction"] = "All pages have been retrieved."

    return [
        TextContent(type="text", text=chunk),
        TextContent(type="text", text=f"\n---\n```json\n{json.dumps(meta, indent=2)}\n```"),
    ]


def _split_into_chunks(text: str, max_tokens: int, counter: Callable[[str], int]) -> list[str]:
    return (
        _try_json_array_split(text, max_tokens, counter)
        or _try_line_split(text, max_tokens, counter)
        or _char_split(text, max_tokens, counter)
    )


def _try_json_array_split(text: str, max_tokens: int, counter: Callable[[str], int]) -> list[str]:
    """Split a JSON array at record boundaries so every chunk is valid JSON."""
    trimmed = text.strip()
    if not (trimmed.startswith("[") and trimmed.endswith("]")):
        return []
    try:
        items = json.loads(trimmed)
        if not isinstance(items, list) or len(items) <= 1:
            return []
    except Exception:
        return []

    chunks: list[str] = []
    batch: list = []

    for item in items:
        batch.append(item)
        if counter(json.dumps(batch)) > max_tokens and len(batch) > 1:
            batch.pop()
            chunks.append(json.dumps(batch, indent=2))
            batch = [item]

    if batch:
        chunks.append(json.dumps(batch, indent=2))

    return chunks if len(chunks) > 1 else []


def _try_line_split(text: str, max_tokens: int, counter: Callable[[str], int]) -> list[str]:
    """Split at newline boundaries — good for logs, CSV, plain text."""
    if "\n" not in text:
        return []

    lines = text.split("\n")
    chunks: list[str] = []
    current: list[str] = []

    for line in lines:
        current.append(line)
        if counter("\n".join(current)) > max_tokens and len(current) > 1:
            current.pop()
            chunks.append("\n".join(current))
            current = [line]

    if current:
        chunks.append("\n".join(current))

    return chunks if len(chunks) > 1 else []


def _char_split(text: str, max_tokens: int, counter: Callable[[str], int]) -> list[str]:
    """Last-resort: binary-search character split."""
    chunks: list[str] = []
    start = 0
    while start < len(text):
        lo, hi = 1, len(text) - start
        while lo < hi:
            mid = (lo + hi + 1) // 2
            if counter(text[start:start + mid]) <= max_tokens:
                lo = mid
            else:
                hi = mid - 1
        chunks.append(text[start:start + lo])
        start += lo
    return chunks


def _content_to_text(result: Any) -> str:
    if isinstance(result, str):
        return result
    if isinstance(result, list):
        parts = []
        for item in result:
            if hasattr(item, "text"):
                parts.append(item.text)
            elif isinstance(item, str):
                parts.append(item)
            else:
                parts.append(json.dumps(item, default=str))
        return "\n".join(parts)
    return json.dumps(result, default=str)


def _to_content_list(result: Any) -> list[TextContent]:
    """Convert any tool result to list[TextContent] for consistent response format."""
    if isinstance(result, list) and result and hasattr(result[0], "type"):
        return result  # already a content list
    return [TextContent(type="text", text=_content_to_text(result))]


def _emit(
    on_paginate: Callable[[PaginateEvent], None] | None,
    event: PaginateEvent,
) -> None:
    if on_paginate is None:
        return
    try:
        on_paginate(event)
    except Exception:
        pass

File: src/paginate.ts
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";
import { ChunkStore } from "./chunk-store.js";
import { defaultTokenCounter, estimateContentTokens } from "./tokenize.js";
import type { PaginateEvent, PaginateOptions } from "./types.js";

const DEFAULTS = {
  maxTokens: 4000,
  ttlMs: 5 * 60 * 1000,
  pageToolName: "get_next_page",
} as const;

/**
 * Wraps an McpServer with transparent cursor-based pagination.
 * Oversized tool responses are chunked and stored; the LLM retrieves
 * subsequent pages via the injected `get_next_page` tool.
 */
export function paginate(server: McpServer, options: PaginateOptions = {}): McpServer {
  const maxTokens = options.maxTokens ?? DEFAULTS.maxTokens;
  const ttlMs = options.ttlMs ?? DEFAULTS.ttlMs;
  const tokenCounter = options.tokenCounter ?? defaultTokenCounter;
  const pageToolName = options.pageToolName ?? DEFAULTS.pageToolName;
  const onPaginate = options.onPaginate;

  const store = new ChunkStore(ttlMs, options.store, options.signingSecret);

  const originalTool = server.tool.bind(server);

  // Register get_next_page via the ORIGINAL method BEFORE proxying, so its
  // own response is never re-paginated by the wrapper.
  originalTool(
    pageToolName,
    "Retrieve the next page of a paginated tool response.",
    { cursor: z.string().describe("Cursor returned by a previous tool call.") },
    async ({ cursor }) => {
      const result = await store.get(cursor);
      if (!result) {
        emit(onPaginate, { type: "cursor_expired" });
        return {
          content: [
            {
              type: "text" as const,
              text: "Cursor not found or expired. Please re-invoke the original tool.",
            },
          ],
          isError: true,
        };
      }
      const { chunk, nextCursor, pageIndex, totalPages } = result;
      emit(onPaginate, { type: "page_fetched", pageIndex, totalPages, hasMore: nextCursor !== null });
      return buildPageResponse(chunk, nextCursor, pageToolName, pageIndex, totalPages);
    }
  );

  // Override server.tool so every SUBSEQUENT registration gets a paginating handler.
  (server as unknown as Record<string, unknown>).tool = new Proxy(originalTool, {
    apply(target, thisArg, args: unknown[]) {
      const patched = patchArgs(args, store, maxTokens, tokenCounter, pageToolName, onPaginate);
      return Reflect.apply(target, thisArg, patched);
    },
  });

  return server;
}

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function patchArgs(
  args: unknown[],
  store: ChunkStore,
  maxTokens: number,
  tokenCounter: (t: string) => number,
  pageToolName: string,
  onPaginate: ((e: PaginateEvent) => void) | undefined
): unknown[] {
  if (args.length === 0) return args;

  // Tool name is always the first argument across all McpServer.tool() overloads.
  const toolName = typeof args[0] === "string" ? args[0] : "unknown";
  const lastIdx = args.length - 1;
  const originalCb = args[lastIdx];
  if (typeof originalCb !== "function") return args;

  const patched = async (...cbArgs: unknown[]) => {
    const result = await (originalCb as (...a: unknown[]) => Promise<unknown>)(...cbArgs);
    return maybePaginate(result, store, maxTokens, tokenCounter, pageToolName, toolName, onPaginate);
  };

  return [...args.slice(0, lastIdx), patched];
}

async function maybePaginate(
  result: unknown,
  store: ChunkStore,
  maxTokens: number,
  tokenCounter: (t: string) => number,
  pageToolName: string,
  toolName: string,
  onPaginate: ((e: PaginateEvent) => void) | undefined
): Promise<unknown> {
  if (!isToolResult(result)) return result;

  const totalTokens = estimateContentTokens(result.content, tokenCounter);
  if (totalTokens <= maxTokens) return result;

  const fullText = contentToText(result.content);
  const chunks = splitIntoChunks(fullText, maxTokens, tokenCounter);
  const id = await store.save(chunks);

  emit(onPaginate, { type: "chunked", toolName, totalTokens, totalChunks: chunks.length });

  const totalPages = chunks.length;
  const firstChunk = chunks[0] ?? "";
  const nextCursor = totalPages > 1 ? store.createCursor(id, 1) : null;
  return buildPageResponse(firstChunk, nextCursor, pageToolName, 0, totalPages);
}

function buildPageResponse(
  chunk: string,
  nextCursor: string | null,
  pageToolName: string,
  pageIndex: number,
  totalPages: number
) {
  const meta = nextCursor
    ? {
        hasMore: true,
        pageIndex,
        totalPages,
        remainingPages: totalPages - pageIndex - 1,
        nextCursor,
        instruction: `Call \`${pageToolName}\` with nextCursor to get the next page. Repeat until hasMore is false.`,
      }
    : {
        hasMore: false,
        pageIndex,
        totalPages,
        remainingPages: 0,
        instruction: "All pages have been retrieved.",
      };

  return {
    content: [
      { type: "text" as const, text: chunk },
      { type: "text" as const, text: `\n---\n\`\`\`json\n${JSON.stringify(meta, null, 2)}\n\`\`\`` },
    ],
  };
}

function splitIntoChunks(
  text: string,
  maxTokens: number,
  tokenCounter: (t: string) => number
): string[] {
  return (
    tryJsonArraySplit(text, maxTokens, tokenCounter) ??
    tryNestedArraySplit(text, maxTokens, tokenCounter) ??
    tryLineSplit(text, maxTokens, tokenCounter) ??
    charSplit(text, maxTokens, tokenCounter)
  );
}

/** Split a JSON array at record boundaries so chunks are always valid JSON. */
function tryJsonArraySplit(
  text: string,
  maxTokens: number,
  counter: (t: string) => number
): string[] | null {
  const trimmed = text.trim();
  if (!trimmed.startsWith("[") || !trimmed.endsWith("]")) return null;

  let items: unknown[];
  try {
    const parsed = JSON.parse(trimmed) as unknown;
    if (!Array.isArray(parsed) || parsed.length <= 1) return null;
    items = parsed;
  } catch {
    return null;
  }

  const chunks: string[] = [];
  let batch: unknown[] = [];

  for (const item of items) {
    batch.push(item);
    if (counter(JSON.stringify(batch)) > maxTokens && batch.length > 1) {
      batch.pop();
      chunks.push(JSON.stringify(batch, null, 2));
      batch = [item];
    }
  }
  if (batch.length > 0) chunks.push(JSON.stringify(batch, null, 2));

  // Only use JSON splitting if we produced multiple chunks.
  return chunks.length > 1 ? chunks : null;
}

/**
 * Handle wrapped JSON objects like { "results": [...], "moves": [...] }.
 * Finds the largest array field, splits it, and reconstructs the wrapper.
 * Fixes edge case: deeply nested objects (e.g. Pokémon full detail) where
 * the top-level is an object, not a bare array.
 */
function tryNestedArraySplit(
  text: string,
  maxTokens: number,
  counter: (t: string) => number
): string[] | null {
  const trimmed = text.trim();
  if (!trimmed.startsWith("{") || !trimmed.endsWith("}")) return null;

  let obj: Record<string, unknown>;
  try {
    const parsed = JSON.parse(trimmed) as unknown;
    if (typeof parsed !== "object" || parsed === null || Array.isArray(parsed)) return null;
    obj = parsed as Record<string, unknown>;
  } catch {
    return null;
  }

  // Find the largest array field to split on
  let bestKey: string | null = null;
  let bestLen = 0;
  for (const [key, val] of Object.entries(obj)) {
    if (Array.isArray(val) && val.length > bestLen) {
      bestKey = key;
      bestLen = val.length;
    }
  }

  if (!bestKey || bestLen <= 1) return null;

  const items = obj[bestKey] as unknown[];
  const wrapper = { ...obj, [bestKey]: [] as unknown[] };

  // Binary-search how many items fit per chunk with the wrapper overhead
  const wrapperTokens = counter(JSON.stringify(wrapper));
  if (wrapperTokens >= maxTokens) return null; // wrapper alone exceeds budget

  const chunks: string[] = [];
  let batch: unknown[] = [];

  for (const item of items) {
    batch.push(item);
    const candidate = { ...wrapper, [bestKey]: batch };
    if (counter(JSON.stringify(candidate)) > maxTokens && batch.length > 1) {
      batch.pop();
      chunks.push(JSON.stringify({ ...wrapper, [bestKey]: batch }, null, 2));
      batch = [item];
    }
  }
  if (batch.length > 0) {
    chunks.push(JSON.stringify({ ...wrapper, [bestKey]: batch }, null, 2));
  }

  return chunks.length > 1 ? chunks : null;
}

/** Split at newline boundaries — good for logs, CSV, plain text. */
function tryLineSplit(
  text: string,
  maxTokens: number,
  counter: (t: string) => number
): string[] | null {
  if (!text.includes("\n")) return null;

  const lines = text.split("\n");
  const chunks: string[] = [];
  let current: string[] = [];

  for (const line of lines) {
    current.push(line);
    if (counter(current.join("\n")) > maxTokens && current.length > 1) {
      current.pop();
      chunks.push(current.join("\n"));
      current = [line];
    }
  }
  if (current.length > 0) chunks.push(current.join("\n"));

  return chunks.length > 1 ? chunks : null;
}

/** Last-resort: binary-search character split. */
function charSplit(
  text: string,
  maxTokens: number,
  counter: (t: string) => number
): string[] {
  const chunks: string[] = [];
  let start = 0;
  while (start < text.length) {
    let lo = 1;
    let hi = text.length - start;
    while (lo < hi) {
      const mid = Math.ceil((lo + hi) / 2);
      if (counter(text.slice(start, start + mid)) <= maxTokens) lo = mid;
      else hi = mid - 1;
    }
    chunks.push(text.slice(start, start + lo));
    start += lo;
  }
  return chunks;
}

function contentToText(content: unknown): string {
  if (Array.isArray(content)) {
    return content
      .map((item) => {
        if (typeof item === "object" && item !== null && "text" in item) {
          return String((item as Record<string, unknown>).text);
        }
        return JSON.stringify(item);
      })
      .join("\n");
  }
  return JSON.stringify(content);
}

interface ToolResult {
  content: unknown;
  isError?: boolean;
}

function isToolResult(value: unknown): value is ToolResult {
  return typeof value === "object" && value !== null && "content" in value;
}

function emit(
  onPaginate: ((e: PaginateEvent) => void) | undefined,
  event: PaginateEvent
): void {
  try {
    onPaginate?.(event);
  } catch {
    // never let a logging callback crash the pagination pipeline
  }
}
